Sources/NIOHTTP2/HTTP2StreamChannel.swift

```diff
@@ -599,4 +646,61 @@ internal extension HTTP2StreamChannel {
     func receiveParentChannelReadComplete() {
         self.tryToRead()
     }
+
+    func parentChannelWritabilityChanged(newValue: Bool) {
+        // There's a trick here that's worth noting: if the child channel hasn't either sent a frame
+        // or been activated on the network, we don't actually want to change the observable writability.
+        // This is because in this case we really want user code to send a frame as soon as possible to avoid
+        // issues with their stream ID becoming out of date. Once the state transitions we can update
+        // the writability if needed.
+        guard case .changed(newValue: let localValue) = self.writabilityManager.parentWritabilityChanged(newValue) else {
+            return
+        }
+
+        // Ok, the writability changed.
+        switch self.state {
+        case .idle, .localActive:
+            // Do nothing here.
+            return
+        case .remoteActive, .active, .closing, .closingNeverActivated, .closed:
+            self._isWritable.store(localValue)
+            self.pipeline.fireChannelWritabilityChanged()
+        }
+    }
+}
+
+extension HTTP2Frame {
+    /// A shorthand heuristic for how many bytes we assume a frame consumes on the wire.
+    ///
+    /// Here we concern ourselves only with per-stream frames: that is, `HEADERS`, `DATA`,
+    /// `WINDOW_UDPATE`, `RST_STREAM`, and I guess `PRIORITY`. As a simple heuristic we
+    /// hard code fixed lengths for fixed length frames, use a calculated length for
+    /// variable length frames, and just ignore encoded headers because it's not worth doing a better
+    /// job.
+    fileprivate var bufferBytes: Int {
+        let frameHeaderSize = 9
+
+        switch self.payload {
+        case .data(let d):
+            let paddingBytes = d.paddingBytes.map { $0 + 1 } ?? 0
+            return d.data.readableBytes + paddingBytes + frameHeaderSize
```

Comment:
I am a bit concerned that data readableBytes could change over the life-time. Wouldn't it be better to capture it when the frame is created and always return the same ?

Reply:
> I am a bit concerned that data readableBytes could change over the life-time.

Over the lifetime of what?

Reply:
lifetime of the frame... that said I just remembered we are not in java-land here and so it should be fine (as it is a struct).

Reply:
Indeed, and as this is Java land the only risk is that this frame could be lengthen by the HTTP/2 implementation itself. As it cannot, there isn't a risk here, and we should be good.